### ISSUE TITLE
IAM controller - Role, spec.PolicyRefs does not work

### ISSUE THREAD

**Issue by michalschott** [aws-controllers-k8s/community#1610]
**Describe the bug**
When trying to apply Role with `spec.policyRefs` controller keeps spamming about resource not being synced:

```
ack-iam-controller-iam-chart-84d496f96f-2p7ng controller 2023-01-05T13:04:24.158Z	ERROR	controller.role	Reconciler error	{"reconciler group": "iam.services.k8s.aws", "reconciler kind": "Role", "name": "iam-controller-test-role", "namespace": "default", "error": "the referenced resource is not synced yet. resource:Policy, namespace:default, name:iam-controller-test-policy"}
ack-iam-controller-iam-chart-84d496f96f-2p7ng controller sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
ack-iam-controller-iam-chart-84d496f96f-2p7ng controller 	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:266
ack-iam-controller-iam-chart-84d496f96f-2p7ng controller sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
ack-iam-controller-iam-chart-84d496f96f-2p7ng controller 	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:227
```

Using `spec.policies` works.

**Steps to reproduce**

Apply below manifest:

```
apiVersion: iam.services.k8s.aws/v1alpha1
kind: Role
metadata:
  name: iam-controller-test-role
  namespace: default
spec:
  name: iam-controller-test-role
  assumeRolePolicyDocument: >
    {
      "Version":"2012-10-17",
      "Statement": [{
        "Effect":"Allow",
        "Principal": {
          "Service": [
            "ec2.amazonaws.com"
          ]
        },
        "Action": ["sts:AssumeRole"]
      }]
    }
  # policies:
  # - iam-controller-test-policy
  policyRefs:
    - from:
        name: iam-controller-test-policy
---
apiVersion: iam.services.k8s.aws/v1alpha1
kind: Policy
metadata:
  name: iam-controller-test-policy
  namespace: default
spec:
  name: iam-controller-test-policy
  policyDocument: '{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":"s3:ListAllMyBuckets","Resource":"arn:aws:s3:::*"},{"Effect":"Allow","Action":["s3:List*"],"Resource":["*"]}]}'
```

**Environment**

controller version 0.1.1

**Comment by michalschott** [comment]
On the other side, using `spec.policy` user could attach some undesired policies like `arn:aws:iam::aws:policy/PowerUserAccess` to their role - is there something I can do about it?

**Comment by jaypipes** [comment]
> On the other side, using `spec.policy` user could attach some undesired policies like `arn:aws:iam::aws:policy/PowerUserAccess` to their role - is there something I can do about it?

Unfortunately, no, this same constraint exists if you allow normal users to create and attach their own policies to their own user resources via the AWS Console. You should limit (by Kubernetes RBAC) the ability of "normal" users to create/read/update `iam.services.k8s.aws/User` and `iam.services.k8s.aws/Policy` resources in the same way you would limit that AWS user's ability to mutate IAM resources in the AWS Console.

**Comment by jaypipes** [comment]
As for the originally-posted bug, I will have a look into this and see why Policy resources are not getting to a Synced state.

**Comment by jaypipes** [comment]
> > > On the other side, using `spec.policy` user could attach some undesired policies like `arn:aws:iam::aws:policy/PowerUserAccess` to their role - is there something I can do about it?
> > 
> > 
> > Unfortunately, no, this same constraint exists if you allow normal users to create and attach their own policies to their own user resources via the AWS Console. You should limit (by Kubernetes RBAC) the ability of "normal" users to create/read/update `iam.services.k8s.aws/User` and `iam.services.k8s.aws/Policy` resources in the same way you would limit that AWS user's ability to mutate IAM resources in the AWS Console.
> 
> My intention is to let dev teams to manage IRSA by themselves in easy, scalable and application-bundled way but also prevent any possibility of unintentional privilege escalation. By denying access to these CRDs on RBAC level theres no gain at all by using this operator. If I understood correctly, looks like this is somehow AWS limitation?

Unfortunately, yes. This is an AWS limitation. AFAIK, there's no way you can set things up so that you allow a user to give themselves (or another principal/role) only *some* policies but not others. In other words, you cannot allow user Joe to grant themselves `arn:aws:iam::aws:policy/ReadOnlyAccess` but *not* `arn:aws:iam::aws:policy/PowerUserAccess`.

AWS principals (users or roles) either have the ability to attach a policy (any policy) to another principal or they don't. There's no in between there, unfortunately.

> Technically I could use something like OPA to prevent such situations, but this would require to enumerate dangerous policies. Combining this with proper code review (we work gitops way) _might_ convince security team that this is enough.

Yes, you could do the above, for sure.

Otherwise, you will want to have different people handle the configuration of IAM identities/principals.

**Comment by michalschott** [comment]
Any ETA on bugix? Maybe I could help / take a look?

**Comment by jaypipes** [comment]
> Any ETA on bugix? Maybe I could help / take a look?

@michalschott I will be looking into this tomorrow morning.

**Comment by RedbackThomson** [comment]
> Technically I could use something like OPA to prevent such situations, but this would require to enumerate dangerous policies. Combining this with proper code review (we work gitops way) _might_ convince security team that this is enough.

+1 for OPA or Kyverno if you can't block access entirely

**Comment by jaypipes** [comment]
Completed with  aws-controllers-k8s/iam-controller#57. The policyRefs fields for Role, User and Group is working correctly.

**Comment by jaypipes** [comment]
@michalschott can you please paste the output of `kubectl describe policies/<your policy name>` for the referenced Policy resource? Thanks!

**Comment by wjf3121** [comment]
+1 I've encountered the same issue described in the comment above with `v1.2.1`:

The policy was created successfully but the `Conditions` shows `Delayed Late Initialization`.

**Comment by a-hilaly** [comment]
This is now fixed in [iam-controller v1.3.6](https://github.com/aws-controllers-k8s/iam-controller/releases/tag/v1.3.6) - the controller now correctly handles the Description field for Roles and Policies, preventing an infinite requeue caused by missing Description field in Create calls. cc @wjf3121 @michalschott
